File: apps/ios/targets/keyboard/AppGroupBridge.swift
import Foundation

enum BridgeKey: String, CaseIterable {
  case bridgeSchemaVersion
  case keyboardSeen
  case keyboardHasFullAccess
  case keyboardStatusRevision
  case keyboardVerificationRequired
  case shortcutAvailable
  case activeModeId
  case keyboardHapticsEnabled
  case keyboardSoundEnabled
  case keyboardPredictionsEnabled
  case keyboardAutocorrectEnabled
  case keyboardSwipeEnabled
  case keyboardPersonalVocabulary
  case keyboardPersonalVocabularyRevision
  case liveActivityDisplayMode
  case streamingInsertionEnabled
  case swipeDecoderStatus
  case onboardingComplete
  case apiBaseURL
  case apiCredential
  case activeModeSnapshot
  case sessionActive
  case sessionOwner
  case sessionPhase
  case sessionErrorMessage
  case sessionHeartbeat
  case sessionStopRequested
  case sessionRevision
  case recordingRequested
  case requestRevision
  case requestedEntryPoint
  case activeRequestId
  case keyboardRequestId
  case partialTranscript
  case partialTranscriptRequestId
  case partialTranscriptRevision
  case finalResultId
  case finalRequestId
  case finalTranscript
  case finalResultStatus
  case transcriptRevision
  case consumedResultId
  case nativeResultEnvelope
  case nativeResultRevision
  case nativeResultConsumedRevision
}

enum KeyboardStatusNotifier {
  private static let fullAccessName = "studio.peacockery.timbervox.keyboard.full-access"
  private static let restrictedName = "studio.peacockery.timbervox.keyboard.restricted"

  static func post(hasFullAccess: Bool) {
    let name = hasFullAccess ? fullAccessName : restrictedName
    CFNotificationCenterPostNotification(
      CFNotificationCenterGetDarwinNotifyCenter(),
      CFNotificationName(name as CFString),
      nil,
      nil,
      true
    )
  }
}

enum KeyboardBridge {
  static let group = "group.studio.peacockery.timbervox"
  static let schemaVersion = 5

  private static var defaults: UserDefaults? {
    UserDefaults(suiteName: group)
  }

  static func initialize() {
    guard integer(for: .bridgeSchemaVersion) < schemaVersion else { return }
    defaults?.removeObject(forKey: "pendingTranscript")
    remove(.finalResultId)
    remove(.finalRequestId)
    remove(.finalTranscript)
    remove(.finalResultStatus)
    remove(.consumedResultId)
    remove(.sessionActive)
    remove(.sessionOwner)
    remove(.sessionPhase)
    remove(.sessionErrorMessage)
    remove(.sessionHeartbeat)
    remove(.sessionStopRequested)
    remove(.sessionRevision)
    remove(.recordingRequested)
    set(schemaVersion, for: .bridgeSchemaVersion)
    seed(true, for: .keyboardHapticsEnabled)
    seed(true, for: .keyboardSoundEnabled)
    seed(true, for: .keyboardPredictionsEnabled)
    seed(true, for: .keyboardAutocorrectEnabled)
    seed(true, for: .keyboardSwipeEnabled)
    seed("waveform", for: .liveActivityDisplayMode)
    seed(false, for: .streamingInsertionEnabled)
  }

  static func bool(for key: BridgeKey) -> Bool {
    defaults?.bool(forKey: key.rawValue) ?? false
  }

  static func integer(for key: BridgeKey) -> Int {
    defaults?.integer(forKey: key.rawValue) ?? 0
  }

  static func string(for key: BridgeKey) -> String? {
    defaults?.string(forKey: key.rawValue)
  }

  static func set(_ value: Bool, for key: BridgeKey) {
    defaults?.set(value, forKey: key.rawValue)
  }

  static func set(_ value: Int, for key: BridgeKey) {
    defaults?.set(value, forKey: key.rawValue)
  }

  static func set(_ value: String, for key: BridgeKey) {
    defaults?.set(value, forKey: key.rawValue)
  }

  static func remove(_ key: BridgeKey) {
    defaults?.removeObject(forKey: key.rawValue)
  }

  static func synchronize() {
    defaults?.synchronize()
  }

  private static func seed(_ value: Bool, for key: BridgeKey) {
    guard defaults?.object(forKey: key.rawValue) == nil else { return }
    set(value, for: key)
  }

  private static func seed(_ value: String, for key: BridgeKey) {
    guard defaults?.object(forKey: key.rawValue) == nil else { return }
    set(value, for: key)
  }
}

File: apps/ios/src/features/keyboard/app-group-bridge.ts
import { ExtensionStorage } from "@bacons/apple-targets";

const APP_GROUP = "group.com.chiejimofor.timbervox";
const BRIDGE_SCHEMA_VERSION = 3;
const storage = new ExtensionStorage(APP_GROUP);

const bridgeKeys = [
  "bridgeSchemaVersion",
  "keyboardSeen",
  "keyboardHasFullAccess",
  "shortcutAvailable",
  "activeModeId",
  "keyboardHapticsEnabled",
  "keyboardSoundEnabled",
  "keyboardPredictionsEnabled",
  "keyboardAutocorrectEnabled",
  "keyboardSwipeEnabled",
  "onboardingComplete",
  "apiBaseURL",
  "apiCredential",
  "activeModeSnapshot",
  "sessionActive",
  "recordingRequested",
  "requestRevision",
  "requestedEntryPoint",
  "activeRequestId",
  "keyboardRequestId",
  "partialTranscript",
  "finalResultId",
  "finalRequestId",
  "finalTranscript",
  "transcriptRevision",
  "consumedResultId",
  "nativeResultEnvelope",
  "nativeResultRevision",
  "nativeResultConsumedRevision",
] as const;

type BridgeKey = (typeof bridgeKeys)[number];

function initializeAppGroupBridge() {
  // App Shortcuts ship with the installed app and are immediately available.
  // This flag represents that capability, not whether the user has run it yet.
  writeBridgeBoolean("shortcutAvailable", true);
  if (readBridgeNumber("bridgeSchemaVersion") >= BRIDGE_SCHEMA_VERSION) return;
  // V1 results had no request ownership, so promoting one could paste stale text.
  storage.remove("pendingTranscript");
  storage.remove("finalResultId");
  storage.remove("finalRequestId");
  storage.remove("finalTranscript");
  storage.remove("consumedResultId");
  writeBridgeNumber("bridgeSchemaVersion", BRIDGE_SCHEMA_VERSION);
  seedBoolean("keyboardHapticsEnabled", true);
  seedBoolean("keyboardSoundEnabled", true);
  seedBoolean("keyboardPredictionsEnabled", true);
  seedBoolean("keyboardAutocorrectEnabled", true);
  seedBoolean("keyboardSwipeEnabled", true);
}

function readBridgeBoolean(key: BridgeKey) {
  return normalizeBridgeBoolean(storage.get(key));
}

function normalizeBridgeBoolean(value: unknown) {
  if (typeof value === "boolean") return value;
  if (typeof value === "number") return value > 0;
  if (typeof value !== "string") return false;
  const normalized = value.trim().toLowerCase();
  return normalized === "true" || normalized === "yes" || normalized === "1";
}

function readBridgeNumber(key: BridgeKey) {
  return normalizeBridgeNumber(storage.get(key));
}

function normalizeBridgeNumber(value: unknown) {
  if (value === true || value === "true" || value === "yes") return 1;
  if (value === false || value === "false" || value === "no") return 0;
  const number = Number(value ?? 0);
  return Number.isFinite(number) ? number : 0;
}

function readBridgeString(key: BridgeKey) {
  const value = storage.get(key);
  return typeof value === "string" ? value : "";
}

function writeBridgeBoolean(key: BridgeKey, value: boolean) {
  storage.set(key, value ? 1 : 0);
}

function writeBridgeNumber(key: BridgeKey, value: number) {
  storage.set(key, value);
}

function writeBridgeString(key: BridgeKey, value: string) {
  storage.set(key, value);
}

function removeBridgeValue(key: BridgeKey) {
  storage.remove(key);
}

function seedBoolean(key: BridgeKey, value: boolean) {
  if (storage.get(key) === null || storage.get(key) === undefined) {
    writeBridgeBoolean(key, value);
  }
}

export {
  APP_GROUP,
  BRIDGE_SCHEMA_VERSION,
  bridgeKeys,
  initializeAppGroupBridge,
  normalizeBridgeBoolean,
  normalizeBridgeNumber,
  readBridgeBoolean,
  readBridgeNumber,
  readBridgeString,
  removeBridgeValue,
  writeBridgeBoolean,
  writeBridgeNumber,
  writeBridgeString,
};
export type { BridgeKey };
